### PR TITLE
When shard becomes active again, immediately increase its indexing buffer

### PR DESCRIPTION
**Pull request by mikemccand**
Spinoff from #13802.

Today, when an index goes from inactive to active, because indexing ops suddenly arrive to an inactive index, we (IndexingMemoryController) take up to 30 seconds to notice this and then increase the indexing buffer from the tiny idle 512KB to "its fair share".  This is somewhat dangerous because it could write many, many segments in those 30 seconds...

This PR changes that so the inactive -> active transition instead causes us to immediately re-visit the indexing buffer for all shards.

It also shifts responsibility of tracking active/inactive into `IndexShard`, and no longer uses the translog ID/ops count to check for changes (I think this is more error prone?  E.g. #13802), simplifying it instead to timestamp (System.nanoTime) of the last indexing op.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -975,6 +999,7 @@ public void addFailedEngineListener(Engine.FailedEngineListener failedEngineList
         this.failedEngineListener.delegates.add(failedEngineListener);
     }
 
+    /** Returns true if the indexing buffer size did change */
     public void updateBufferSize(ByteSizeValue shardIndexingBufferSize, ByteSizeValue shardTranslogBufferSize) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I don't think this doc is correct - the return type is void

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java [hunk 2]
```diff
@@ -255,49 +251,43 @@ protected void updateShardBuffers(ShardId shardId, ByteSizeValue shardIndexingBu
...
+    protected boolean isShardInactive(ShardId shardId, long inactiveTimeNS) {
+        final IndexShard shard = getShard(shardId);
+        if (shard == null) {
+            return false;
+        }
+        return currentTimeInNanos() - shard.getLastWriteNS() >= inactiveTimeNS;
+    }
+
 
     /** returns the current translog status (generation id + ops) for the given shard id. Returns null if unavailable. */
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: Out of date comment?

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java [hunk 3]
```diff
@@ -255,49 +251,43 @@ protected void updateShardBuffers(ShardId shardId, ByteSizeValue shardIndexingBu
         }
     }
 
+    protected boolean isShardInactive(ShardId shardId, long inactiveTimeNS) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: Would `isShardActive` be simpler to understand just because it tests a positive?

Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: Good point about using negative name ... I think I'll rename to `isShardIdle`?

**Comment by mikemccand** [comment]
Thanks @nik9000 I folded in the feedback.

File: core/src/main/java/org/elasticsearch/index/engine/ShadowEngine.java [hunk 4]
```diff
@@ -245,4 +245,9 @@ protected SegmentInfos getLastCommittedSegmentInfos() {
         return lastCommittedSegmentInfos;
     }
 
+    @Override
+    public long indexWriterRAMBytesUsed() {
+        // No IndexWriter
+        return 0L;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/ShadowEngine.java]: I wonder if UnsupportedOperationException is a better choice here- we don't have an indexwriter...

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 5]
```diff
@@ -190,6 +192,10 @@
 
     private final IndexSearcherWrapper searcherWrapper;
 
+    private final AtomicBoolean active = new AtomicBoolean();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: can we have some java docs here?

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 6]
```diff
@@ -914,7 +925,20 @@ public void readAllowed() throws IllegalIndexShardStateException {
...
 
-    private void writeAllowed(Engine.Operation.Origin origin) throws IllegalIndexShardStateException {
+    /** Returns timestamp of last indexing operation */
+    public long getLastWriteNS() {
+        return lastWriteNS;
+    }
+
+    private void ensureWriteAllowed(Engine.Operation op) throws IllegalIndexShardStateException {
+        lastWriteNS = op.startTime();
+        if (active.getAndSet(true) == false) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: can we do this once we check the write is indeed allowed? Also I think it will be clearer if we have this in a dedicated method (`markLastWrite` or something like that)

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 7]
```diff
@@ -242,11 +248,16 @@ public IndexShard(ShardId shardId, @IndexSettings Settings indexSettings, ShardP
...
         this.indexShardOperationCounter = new IndexShardOperationCounter(logger, shardId);
+        this.indexingMemoryController = provider.getIndexingMemoryController();
+
         this.searcherWrapper = provider.getIndexSearcherWrapper();
         this.percolatorQueriesRegistry = new PercolatorQueriesRegistry(shardId, indexSettings, queryParserService, indexingService, mapperService, indexFieldDataService);
         if (mapperService.hasMapping(PercolatorService.TYPE_NAME)) {
             percolatorQueriesRegistry.enableRealTimePercolator();
         }
+
+        // TODO: can we somehow call IMC.forceCheck here?  Since we just became active, it can divvy up the RAM
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I assume you tried? what broke?

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 8]
```diff
@@ -993,28 +1019,40 @@ public void updateBufferSize(ByteSizeValue shardIndexingBufferSize, ByteSizeValu
...
+                // so it clears its buffers (otherwise it won't clear until the next indexing/delete op)
+                logger.debug("refresh because index buffer decreased to [{}] and IndexWriter is now using [{}] bytes",
+                             shardIndexingBufferSize, iwBytesUsed);
 
-            if (shardIndexingBufferSize == EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER) {
-                // it's inactive: make sure we do a refresh / full IW flush in this case, since the memory
-                // changes only after a "data" change has happened to the writer
-                // the index writer lazily allocates memory and a refresh will clean it all up.
-                logger.debug("updating index_buffer_size from [{}] to (inactive) [{}]", preValue, shardIndexingBufferSize);
+                // TODO: should IW have an API to move segments to disk, but not refresh?
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I'm curious about this one - I thought the segments are always (once made) on disk - but not fsynced?

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 9]
```diff
@@ -993,28 +1019,40 @@ public void updateBufferSize(ByteSizeValue shardIndexingBufferSize, ByteSizeValu
         if (preValue.bytes() != shardIndexingBufferSize.bytes()) {
             // so we push changes these changes down to IndexWriter:
             engine.onSettingsChanged();
+            logger.debug("updating index_buffer_size from [{}] to [{}]", preValue, shardIndexingBufferSize);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: can we make this log include the decision about whether to refresh (and also the current IW memory consumption, while at it)

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 10]
```diff
@@ -993,28 +1019,40 @@ public void updateBufferSize(ByteSizeValue shardIndexingBufferSize, ByteSizeValu
...
 
         engine.getTranslog().updateBuffer(shardTranslogBufferSize);
     }
 
+    /** Record that this shard is now inactive, and decrease the indexing and translog buffers to tiny values. */
     public void markAsInactive() {
-        updateBufferSize(EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER, TranslogConfig.INACTIVE_SHARD_TRANSLOG_BUFFER);
-        indicesLifecycle.onShardInactive(this);
+        if (active.getAndSet(false)) {
+            updateBufferSize(EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER, TranslogConfig.INACTIVE_SHARD_TRANSLOG_BUFFER);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I know it's not part of this change, but it bugs me (for a while now) that the indexing memory controller owns the buffer size for active indices but inactivity is controlled by the engine/shard. I wonder if we should move these settings to the memory controller.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 11]
```diff
@@ -993,28 +1019,40 @@ public void updateBufferSize(ByteSizeValue shardIndexingBufferSize, ByteSizeValu
...
+            logger.debug("shard is now inactive");
+            indicesLifecycle.onShardInactive(this);
+        }
+    }
+
+    /** Returns {@code true} if this shard is active (has seen indexing ops in the last {@link
+     *  IndexingMemoryController#SHARD_INACTIVE_TIME_SETTING} (default 5 minutes), else {@code false}. */
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I wonder if it now makes more sense to have this setting on a shard level (since shard owns the inactivity). We can fall back to the old, node level settings (or just remove it if this is 3.0 change only and deprecate on 2.x)

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java [hunk 12]
```diff
@@ -255,49 +250,45 @@ protected void updateShardBuffers(ShardId shardId, ByteSizeValue shardIndexingBu
         }
     }
 
+    /** returns {@code Boolean.TRUE} if no indexing operations have arrived to this shard since the {@code inactiveTimeNS},
+     *  {@code Boolean.FALSE} if they have, and {@code null} if the shard is unknown */
+    protected boolean isShardIdle(ShardId shardId, long inactiveTimeNS) {
+        final IndexShard shard = getShard(shardId);
+        if (shard == null) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: this isn't what the comment says? (return null)

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java [hunk 13]
```diff
@@ -255,49 +250,45 @@ protected void updateShardBuffers(ShardId shardId, ByteSizeValue shardIndexingBu
...
 
     class ShardsIndicesStatusChecker implements Runnable {
 
-        private final Map<ShardId, ShardIndexingStatus> shardsIndicesStatus = new HashMap<>();
+        // True if the shard was active last time we checked
+        private final Map<ShardId,Boolean> shardWasActive = new HashMap<>();
 
         @Override
-        public void run() {
+        public synchronized void run() {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: good catch!

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java [hunk 14]
```diff
@@ -255,49 +250,45 @@ protected void updateShardBuffers(ShardId shardId, ByteSizeValue shardIndexingBu
...
     }
 
+    /** returns {@code Boolean.TRUE} if no indexing operations have arrived to this shard since the {@code inactiveTimeNS},
+     *  {@code Boolean.FALSE} if they have, and {@code null} if the shard is unknown */
+    protected boolean isShardIdle(ShardId shardId, long inactiveTimeNS) {
+        final IndexShard shard = getShard(shardId);
+        if (shard == null) {
+            return false;
+        }
+        return currentTimeInNanos() - shard.getLastWriteNS() >= inactiveTimeNS;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: Shall we push this into IndexShard as well? have a method like

```
IndexShard.checkIsIdle() {
   if (System.currentTime - lastWriteNs() > inactiveTimeNs) {
         active.set(false);
  }
  return active.get() == false;
}
```

so all inactive/active logic is in IndexShard and you just ask for a check from the outside.

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java [hunk 15]
```diff
@@ -364,12 +360,15 @@ private int updateShardStatuses(EnumSet<ShardStatusChangeType> changes, List<Sha
             return changes;
         }
 
-        private void calcAndSetShardBuffers(int activeShards, String reason) {
-            if (activeShards == 0) {
+        private void calcAndSetShardBuffers(int activeShardCount, String reason) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: I think things will be simpler if this method is driven by the current shardWasActive map. This means we don't need to count activeShardCount and also not worry about whether availableShards() changed it's mind (see loop at the end of this method)

File: core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerTests.java [hunk 16]
```diff
@@ -39,44 +43,28 @@
...
+        final Map<ShardId, Long> lastIndexTimeNanos = new HashMap<>();
+        final Set<ShardId> activeShards = new HashSet<>();
+
         long currentTimeSec = TimeValue.timeValueNanos(System.nanoTime()).seconds();
 
         public MockController(Settings settings) {
             super(Settings.builder()
                             .put(SHARD_INACTIVE_INTERVAL_TIME_SETTING, "200h") // disable it
-                            .put(SHARD_INACTIVE_TIME_SETTING, "0s") // immediate
+                            .put(SHARD_INACTIVE_TIME_SETTING, "1ms") // nearly immediate
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerTests.java]: wondering - why did you change this?

File: core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerTests.java [hunk 17]
```diff
@@ -139,13 +133,13 @@ public void testShardAdditionAndRemoval() {
...
         final ShardId shard1 = new ShardId("test", 1);
-        controller.setTranslog(shard1, randomInt(10), randomInt(10));
+        controller.simulateIndexing(shard1);
         controller.forceCheck();
         controller.assertBuffers(shard1, new ByteSizeValue(10, ByteSizeUnit.MB), new ByteSizeValue(64, ByteSizeUnit.KB)); // translog is maxed at 64K
 
         // add another shard
         final ShardId shard2 = new ShardId("test", 2);
-        controller.setTranslog(shard2, randomInt(10), randomInt(10));
+        controller.simulateIndexing(shard2);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerTests.java]: I wonder if simulate indexing should also call forceCheck (like real indexing would do)..

**Comment by bleskes** [comment]
I love how this turned out, especially around moving away from the translog checks. I left some suggestions here and there. I also think we should  move the controller to work on IndexShard objects rather than shard ids (sorry for the scope creep - can be done in a different change). At some point I got worried that a new shard will not be seen as "added" if the original shard was destroyed and a new one created. I think it's OK now (because we only use current state when allocating memory) but it's more correct to use physical instances.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 7 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: > I assume you tried? what broke?

I didn't try!  It just looked not simple because here we are in this `IndexShard` ctor and if we invoke IMC at this point, probably the caller/Guice (creating this shard, IndicesService maybe?) won't even have this shard listed in its state?

Maybe up above somewhere I could call it ... I'll try.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 8 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: > I thought the segments are always (once made) on disk - but not fsynced?

IW writes new segments to disk and only fsyncs once IW.commit (ES flush) is called.

But from ES the only way we have to get IW to move in-memory segments to disk is via refresh, but that's overkill (it writes the segment and also opens new `SegmentReader`) and if the ES user e.g. has high or disabled `refresh_interval` this is wasted work.

IW doesn't give a public method to do this today ... I suppose we could `addIndexes` with nothing :)

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 9 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: > can we make this log include the decision about whether to refresh (and also the current IW memory consumption, while at it)

You mean have a single debug log line instead of two?  OK ...

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 10 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Good idea, I'll move both translog and indexing inactive buffer constants to IMC.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 9 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: meh - missed the addition of the other one (just saw the removal of the log in the if), but yeah - one is nicer..

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 11 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: > I wonder if it now makes more sense to have this setting on a shard level (since shard owns the inactivity). 

So users could change the inactive time per shard instead of globally for all shards on the node?

I don't think that's a good idea:)  I think there are too many knobs already!  Taking one knob and turning it into 20 seems not great...

File: core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerTests.java (hunk 16 above)
Comment (mikemccand) [inline core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerTests.java]: > wondering - why did you change this?

Oh it's because I changed the inactive time setting to be >= not >, i.e. an inactive time of 60s means "if you've been idle for 60s or more, mark yourself inactive".

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java (hunk 14 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: This is a good idea!  I think I'll take it further and fold it into a single method that checks for idle, and becomes idle (dropping the buffers) if so.

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java (hunk 15 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: Hmm, confused: seems like we still need `activeShardCount` up front since we then divide total allowed heap by that?

I agree the loop at the end of the method should just use shardWasActive ... I'll fix!

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: > Hmm, confused: seems like we still need activeShardCount up front since we then divide total allowed heap by that?

what I mean that we should just loop on shardWasActive twice here, once to calculate activeShardCount and one to apply it. I think it will be clearer than passing it as an argument and trying to make sure it's correct in another functionn

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 11 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: fare enough. It would be a settings per index, but I get your point.

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java (hunk 15 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: Ahh gotchya, will do!

**Comment by mikemccand** [comment]
> I also think we should move the controller to work on IndexShard objects rather than shard ids

++

But I thought we used `ShardId` all over to make this more easily unit-tested?  But anyway let's do this on a separate change ... this one is big already :)

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 7 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: OK I added a line at the end of `IndexService.createShard` to invoke `forceCheck` ... but are we worried about the O(N^2) cost here?  N should be smallish (number of nodes on the shard)?

**Comment by mikemccand** [comment]
OK I folded feedback @bleskes, thank you!

File: core/src/main/java/org/elasticsearch/index/IndexService.java [hunk 18]
```diff
@@ -288,6 +288,7 @@ public synchronized IndexShard createShard(int sShardId, ShardRouting routing) {
             indicesLifecycle.afterIndexShardCreated(indexShard);
             settingsService.addListener(indexShard);
             shards = newMapBuilder(shards).put(shardId.id(), indexShard).immutableMap();
+            indexServicesProvider.getIndexingMemoryController().forceCheck();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: I was thinking about it and I think we should start the shard as inactive (it is :)) and only mark it as active + the claiming of memory in IndexShard#skipTranslogRecovery / IndexShard#performTranslogRecovery when we actually open the engine for indexing. We also make sure call IndexingMemoryController.forceCheck before we create the engine, so we always have an up to date buffer size ... 

Re concerns about n^2 problem, it is there but since we limit the amount of concurrent recoveries a node can do, it means it will be done during a longer process. Also because of the limit it is hard to do anything around batching. All in all, I think we're OK.

File: core/src/main/java/org/elasticsearch/index/engine/EngineConfig.java [hunk 19]
```diff
@@ -108,7 +108,6 @@
     public static final TimeValue DEFAULT_REFRESH_INTERVAL = new TimeValue(1, TimeUnit.SECONDS);
     public static final TimeValue DEFAULT_GC_DELETES = TimeValue.timeValueSeconds(60);
     public static final ByteSizeValue DEFAULT_INDEX_BUFFER_SIZE = new ByteSizeValue(64, ByteSizeUnit.MB);
-    public static final ByteSizeValue INACTIVE_SHARD_INDEXING_BUFFER = ByteSizeValue.parseBytesSizeValue("500kb", "INACTIVE_SHARD_INDEXING_BUFFER");
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/EngineConfig.java]: awesome

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 20]
```diff
@@ -994,27 +1031,48 @@ public void updateBufferSize(ByteSizeValue shardIndexingBufferSize, ByteSizeValu
...
+     *  if the shard did in fact become inactive, else false. */
+    public boolean checkIdle(long inactiveTimeNS) {
+        if (System.nanoTime() - lastWriteNS >= inactiveTimeNS && active.getAndSet(false)) {
+            updateBufferSize(IndexingMemoryController.INACTIVE_SHARD_INDEXING_BUFFER, IndexingMemoryController.INACTIVE_SHARD_TRANSLOG_BUFFER);
+            logger.debug("shard is now inactive");
+            indicesLifecycle.onShardInactive(this);
+            return true;
+        }
+
+        return false;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I think we should return active.get() == false and also - set it if we became idle...

File: core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerTests.java (hunk 16 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerTests.java]: OK. I'll brace for the beast which will break this because it runs faster than 1ms :)

File: core/src/main/java/org/elasticsearch/index/IndexService.java (hunk 18 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: Actually, why not simply start up as inactive, init the buffers to the INACTIVE values, and then the first indexing op that arrives will make it active like normal?

OK that's a good point, that concurrent recoveries are limited anyway, so the O(N^2) cost is paid over a long time.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 21]
```diff
@@ -994,27 +1031,48 @@ public void updateBufferSize(ByteSizeValue shardIndexingBufferSize, ByteSizeValu
...
-                logger.debug("updating index_buffer_size from [{}] to (inactive) [{}]", preValue, shardIndexingBufferSize);
+            long iwBytesUsed = engine.indexWriterRAMBytesUsed();
+
+            String message = LoggerMessageFormat.format("updating index_buffer_size from [{}] to [{}]; IndexWriter now using [{}] bytes",
+                                                        preValue, shardIndexingBufferSize, iwBytesUsed);
+
+            if (iwBytesUsed > shardIndexingBufferSize.bytes()) {
+                // our allowed buffer was changed to less than we are currently using; we ask IW to refresh
+                // so it clears its buffers (otherwise it won't clear until the next indexing/delete op)
+                logger.debug(message + "; now refresh to clear IndexWriter memory");
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I'm used to wrapping debug logging stuff in `if (logger.isDebugEnabled())` tests to prevent the message construction in the (very common) case that debug is disabled. Here it probably doesn't matter because message construction is dwarfed by the refresh call.

File: core/src/main/java/org/elasticsearch/index/IndexService.java (hunk 18 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: > Actually, why not simply start up as inactive, init the buffers to the INACTIVE values, and then the first indexing op that arrives will make it active like normal?

That was my original thought as well, but then I remembered you were very worried about indexing into a small buffer. When one creates an index by indexing into it that index will likely get many indexing requests at once, the first will trigger a forced check and the others will fly through into the engine. Figured we better not have to worry about it but If you're good with that, I'm good :)

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 22]
```diff
@@ -994,27 +1031,48 @@ public void updateBufferSize(ByteSizeValue shardIndexingBufferSize, ByteSizeValu
...
     }
 
-    public void markAsInactive() {
-        updateBufferSize(EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER, TranslogConfig.INACTIVE_SHARD_TRANSLOG_BUFFER);
-        indicesLifecycle.onShardInactive(this);
+    /** Called by {@link IndexingMemoryController} to check whether more than {@code inactiveTimeNS} has passed since the last
+     *  indexing operation, and become inactive (reducing indexing and translog buffers to tiny values) if so.  This returns true
+     *  if the shard did in fact become inactive, else false. */
+    public boolean checkIdle(long inactiveTimeNS) {
+        if (System.nanoTime() - lastWriteNS >= inactiveTimeNS && active.getAndSet(false)) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Can you move the getAndSet to its own if statement? It has a side effect and its mixed with stuff that doesn't and when I first read the code I didn't notice it.

File: core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerTests.java (hunk 16 above)
Comment (mikemccand) [inline core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerTests.java]: Wait, this test does not actually rely on wall clock time :)  We advance the clock on our own, deterministically ... the fastest beast in the world should not break it!

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java [hunk 23]
```diff
@@ -364,12 +346,25 @@ private int updateShardStatuses(EnumSet<ShardStatusChangeType> changes, List<Sha
...
+            int activeShardCount = 0;
+            for (Map.Entry<ShardId,Boolean> ent : shardWasActive.entrySet()) {
+                if (ent.getValue()) {
+                    activeShardCount++;
+                }
+            }
+
+            // TODO: we could be smarter here by taking into account how RAM the IndexWriter on each shard
+            // is actually using (using IW.ramBytesUsed), so that small indices (e.g. Marvel) would not
+            // get the same indexing buffer as large indices.  But it quickly gets tricky...
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: Indeed! This sounds like a fun thing to think about!

File: core/src/main/java/org/elasticsearch/index/IndexService.java (hunk 18 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: > If you're good with that, I'm good :)

I'm gonna try :)

File: core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java (hunk 23 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java]: > Indeed! This sounds like a fun thing to think about!

I'll open a followon issue for this ... I think often there are very lopsided shards in a single node, and evenly dividing the indexing buffer seems crazy.

It would be better if the whole system acted like IW acts with N threads using up heap: as soon as the sum of all heap used by these threads exceeds the global budget, we ask the largest one to write to disk.

**Comment by mikemccand** [comment]
OK thanks @nik9000 and @bleskes, I folded in the last round of feedback.

File: core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerTests.java (hunk 16 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indices/memory/IndexingMemoryControllerTests.java]: Oei, I forgot about the virtual time I added - smart move :)

**Comment by bleskes** [comment]
LGTM. Awesome mike.
